src/stdune/path.ml

```diff
@@ -221,86 +230,189 @@ module Local = struct
     loop (to_list t) (to_list from)
 end
 
-type t = string
-let compare = String.compare
+module Kind = struct
+  type t =
+    | External of External.t
+    | Local    of Local.t
 
-module Set = struct
-  include String.Set
-  let sexp_of_t t = Sexp.To_sexp.(list string) (String.Set.to_list t)
-  let of_string_set = map
+  let to_absolute_filename t ~root =
+    match t with
+    | External s -> s
+    | Local l -> Filename.concat root (Local.to_string l)
+
+  let to_string = function
+    | Local "" -> "."
+    | Local t
+    | External t -> t
+
+  let sexp_of_t t = Sexp.atom_or_quoted_string (to_string t)
+
+  let of_string s =
+    if s = "" || Filename.is_relative s then
+      Local s
+    else
+      External s
+
+  let relative ?error_loc t fn =
+    match t with
+    | Local t -> Local (Local.relative ?error_loc t fn)
+    | External t -> External (External.relative t fn)
+
+  let mkdir_p = function
+    | Local t -> Local.mkdir_p t
+    | External t -> External.mkdir_p t
 end
 
-module Map = String.Map
+let (build_dir_kind, set_build_dir) =
+  let build_dir = ref None in
+  let set_build_dir new_build_dir =
+    match !build_dir with
+    | None -> build_dir := Some new_build_dir
+    | Some build_dir ->
+      Exn.code_error "set_build_dir: cannot set build_dir more than once"
+        [ "build_dir", Kind.sexp_of_t build_dir
+        ; "New_build_dir", Kind.sexp_of_t new_build_dir ]
+  in
+  let build_dir () =
+    match !build_dir with
+    | None ->
+      Exn.code_error "build_dir: cannot use build dir before it's set" []
+    | Some build_dir -> build_dir in
+  (build_dir, set_build_dir)
+
+let drop_build_dir p =
+  match build_dir_kind (), Kind.of_string p with
+  | Kind.External bd, Kind.External p ->
+    let open Option.O in
```

Comment:
This seems a bit fragile to me. Because we don't normalize external paths, we might miss cases here. Why do we need this exactly?

Reply:
If the user does something like `$ jbuilder build --build-dir /foo/bar /foo/bar/baz` I'd expect the path to be resolved to a build dir.

I also think that the db file also relies on this. Though we can fix that by changing the sexp reading/writing not to erase the the constructor of `Path.t`.

Reply:
Yes, but what if the the user does `jbuilder build --build-dir /foo//bar /foo/bar/baz`?